module/zfs/spa.c

```diff
@@ -5857,10 +5866,15 @@ spa_vdev_add(spa_t *spa, nvlist_t *nvroot)
 	 */
 	(void) spa_vdev_exit(spa, vd, txg, 0);
 
-	mutex_enter(&spa_namespace_lock);
+	/*
+	 * At this point we just need to serialize this update
+	 * within the pool and the only time we need to take the
+	 * namespace_lock is when we update the global cache
+	 */
+	mutex_enter(&spa->spa_config_update_lock);
```

Comment:
Hmm, it looks like there's an old bug here which could result in a panic. The problem is that once we call `spa_vdev_exit` we are no long guaranteed that the `spa` can't go away. We really should have grabbed a reference on the `spa` to prior to calling `spa_vdev_exit`. Your proposed code change could suffer from the same problem.

Reply:
You are right. I did notice this bug but somehow forgot to take care of it.

Reply:
Actually, I remember now, spa_vdev_add is called from zfs_ioc_vdev_add which before calling spa_vdev_add takes a reference to spa hence there is no need to take reference after spa_vdev_exit

Reply:
Great, thanks for tracking that down.